src/librustc_target/spec/apple_base.rs

```diff
@@ -51,3 +51,63 @@ pub fn macos_llvm_target(arch: &str) -> String {
     let (major, minor) = macos_deployment_target();
     format!("{}-apple-macosx{}.{}.0", arch, major, minor)
 }
+
+#[cfg(target_os = "macos")]
+pub fn sysroot(sdk: &str) -> Result<Option<String>, String> {
+    // Like Clang, allow the `SDKROOT` environment variable used by Xcode to define the sysroot.
+    if let Some(sdk_root) = env::var("SDKROOT").ok() {
+        let actual_sdk_path = sdk_path(sdk)?;
+        let sdk_root_p = Path::new(&sdk_root);
+        // Ignore `SDKROOT` if it's not a valid path.
+        if !sdk_root_p.is_absolute() || sdk_root_p == Path::new("/") || !sdk_root_p.exists() {
+            return Ok(Some(actual_sdk_path));
+        }
+        // Ignore `SDKROOT` if it's clearly set for the wrong platform, which may occur when we're
+        // compiling a custom build script while targeting iOS for example.
+        return Ok(Some(match sdk {
+            "iphoneos" if sdk_root.contains("iPhoneSimulator.platform")
+                || sdk_root.contains("MacOSX.platform") => actual_sdk_path,
```

Comment:
I'm sort of surprised by this logic, I'd sort of suspect something along the lines of:

* If `SDKROOT` is set, only return it if the actual compiler target looks like it's part of `SDKROOT`
* Otherwise, for ios/etc platforms, use `xcrun` to auto-find a root.
* Otherwise if SDKROOT is set and the target is `{x86_64,i686}-apple-darwin`, ensure that `SDKROOT` is removed from linker commands
* Otherwise, return `None`.

I think this would assuage my concern below about passing `SDKROOT` arguments to all rustc invocations for x86_64/i686 by default. That seems susceptible to breakage I fear. Additionally `xcrun` isn't executed if `SDKROOT` is configured, and for cross compilations we'd scrub `SDKROOT` from the environment when executing the linker for the host.